### ISSUE TITLE
no-unbound-method: referencing methods that are bound in constructor should be ok

### ISSUE THREAD

**Issue by unstubbable** [palantir/tslint#3016]
### Bug Report

In https://github.com/palantir/tslint-react/issues/63 it was mentioned that a rule such as `no-unbound-method` should check for methods that are bound in the constructor (e.g. `this.handleClick = this.handleClick.bind(this);`) to avoid false positives. Although https://github.com/palantir/tslint-react/issues/63 was closed with the [introduction of `no-unbound-method`](#2089) this case does not seem to be implemented.

- __TSLint version__: 5.5.0
- __TypeScript version__: 2.4.1
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
class C {
  public constructor() {
    this.method = this.method.bind(this);
  }
  public method(): number {
    return 1;
  }
}

const c = new C();
[0].forEach(c.method);
```

with `tslint.json` configuration:

```json
{
  "no-unbound-method": true
}
```

#### Actual behavior

⚠️ Fails with:

```
[11, 13]: Avoid referencing unbound methods which may cause unintentional scoping of 'this'.
```

#### Expected behavior

✅  Has no errors.

**Comment by adidahiya** [comment]
I guess your suggestion is technically valid, but I'm not sure if the the extra complexity in this rule is worth it -- since this is TS, couldn't you just bind your methods using arrow functions?

**Comment by KingHenne** [comment]
I guess arrow functions as class properties would be fine for most cases in the React world (referring to https://github.com/palantir/tslint-react/issues/63). But they do have the disadvantage that you can not call super in them, i.e. something like this is not possible without the `bind` in the constructor as far as I know:

```ts
class A {
  count: number;

  constructor() {
    this.count = 0;
    this.method = this.method.bind(this);
  }

  method() {
    this.count++;
  }
}

class B extends A {
  method() {
    super.method();
    this.count++;
    console.log(this.count);
  }
}

const b = new B();
// This would be an event handler or something:
[0].forEach(b.method); // logs 2
```

**Comment by KingHenne** [comment]
closing, see https://github.com/palantir/tslint/pull/3017#issuecomment-314368510

**Comment by Amaimersion** [comment]
+1 for re-open. It would be very good if someone will implement checking for binding in the constructor. There was a try (#3017), but it has failed. `no-unbound-method` rule is really a pain for React, because recommend way is to bind in the constructor, not to create arrow lambda.

**Comment by JoshuaKGoldberg** [comment]
You can refile a new issue I'd you'd lilezz so you have authorship of it.

**Comment by karanrajpal** [comment]
Is this being tracked in any open issue? This rule will be really useful if it checks if the function is bound in the constructor.
I converted a bunch of class functions to be arrow functions, but that makes the class not mockable.

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖  #4534 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
